### PR TITLE
Feature: Migrate from panache active record to repository pattern

### PR DESCRIPTION
**Pull request by infeo**
This PR contains the migration from the panache [active record](https://quarkus.io/guides/hibernate-orm-panache#solution-1-using-the-active-record-pattern) to the [repository pattern](https://quarkus.io/guides/hibernate-orm-panache#solution-2-using-the-repository-pattern).

Reasons for this migration:
* improved testability: Less static mocks or PanacheMocking, you can simply mock the repository with regular mockito methods
* less byte code manipulation: Easiert debugging and less arcane bugs due on-the-fly bytecode generation.

Regarding the implemenation, no entitiy extends the PanacheEntityBase anymore, but are plain POJOs with Jakarta persistence annotations. The panache repositories are static subclasses of the enitites.

Additional changes:
* moved audit entities to own package
* factor the license IT tests to a single non-quarkus unit test
* license holder refactoring for improved test coverage

### CODE REVIEW

File: backend/src/test/java/org/cryptomator/hub/license/LicenseValidatorTest.java [hunk 1]
```diff
@@ -71,25 +71,4 @@ public void testValidateMalformedToken() {
...
-	}
-
-	@Test
-	@DisplayName("validate expired token's refreshURL with invalid signature")
-	public void testInvalidSignatureTokensRefreshUrl() {
-		Assertions.assertThrows(SignatureVerificationException.class, () -> {
-			validator.refreshUrl(TOKEN_WITH_INVALID_SIGNATURE);
-		});
-	}
-
```
Comment (overheadhunter) [inline backend/src/test/java/org/cryptomator/hub/license/LicenseValidatorTest.java:94]: why have these tests been removed?

**Review by overheadhunter — COMMENTED** [review]
Looks good, however I have a few specific (see comments on code) and two general questions:

1. Can we make the entities' fields private? Or what is the rationale of making them just package-private?
2. Shouldn't we at least for the `*Event` type entities try to use a single repository (`eventRepo.persist(new SpecificEvent(foo, bar, baz))`)? I guess this would reduce a lot of boilerplate code

File: backend/src/test/java/org/cryptomator/hub/license/LicenseValidatorTest.java (hunk 1 above)
Comment (infeo) [reply backend/src/test/java/org/cryptomator/hub/license/LicenseValidatorTest.java:94]: because the method `LicenseValidator::refreshUrl` does not exist anymore.

File: backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java [hunk 2]
```diff
@@ -95,62 +98,68 @@ void applyInitialHubIdAndLicense(String initialId, String initialLicense) {
...
-	@Scheduled(cron = "0 0 1 * * ?", timeZone = "UTC", concurrentExecution = Scheduled.ConcurrentExecution.SKIP)
-	void refreshLicenseScheduler() throws InterruptedException {
-		if (license != null) {
-			randomMinuteSleeper.sleep(); // add random sleep between [0,59]min to reduce infrastructure load
-			var refreshUrl = licenseValidator.refreshUrl(license.getToken());
-			if (refreshUrl.isPresent()) {
-				var client = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
-				refreshLicense(refreshUrl.get(), license.getToken(), client);
+	@Scheduled(cron = "0 0 1 * * ?", timeZone = "UTC", concurrentExecution = Scheduled.ConcurrentExecution.SKIP, skipExecutionIf = LicenseHolder.class)
+	@NonBlocking
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java:111]: I just read the javadoc again and noticed that this is just a marker annotation. I thought it actually changes execution.

I will remove it.

File: backend/src/main/resources/application.properties [hunk 3]
```diff
@@ -58,7 +58,6 @@ quarkus.datasource.db-kind=postgresql
 quarkus.datasource.jdbc.driver=org.postgresql.Driver
 quarkus.datasource.jdbc.transaction-requirement=off
 quarkus.datasource.jdbc.max-size=16
-quarkus.hibernate-orm.dialect=org.hibernate.dialect.PostgreSQLDialect
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/resources/application.properties:61]: Moved the change to https://github.com/cryptomator/hub/commit/dcf15877ad0139b8a82c716b1bb9f06e881bbbc5

File: backend/src/main/java/org/cryptomator/hub/entities/Group.java [hunk 4]
```diff
@@ -20,6 +22,17 @@ public class Group extends Authority {
...
+		return members;
+	}
+
+	public void setMembers(Set<Authority> members) {
+		this.members = members;
+	}
+
+	@ApplicationScoped
+	public static class Repository implements PanacheRepositoryBase<Group, String> {
+	}
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/entities/Group.java:37]: Yes. We need the repository, otherwise we cannot query/persist the entity.

File: backend/src/main/resources/application.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/resources/application.properties]: fixed in https://github.com/cryptomator/hub/pull/266/commits/081349e42c5e0920a2df477b77d419b08200e849

File: backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java]: fixed in https://github.com/cryptomator/hub/pull/266/commits/081349e42c5e0920a2df477b77d419b08200e849

File: backend/src/main/java/org/cryptomator/hub/entities/Settings.java [hunk 5]
```diff
@@ -48,8 +73,11 @@ public int hashCode() {
...
 	}
 
-	public static Settings get() {
-		return Objects.requireNonNull(Settings.findById(SINGLETON_ID), "Settings not initialized");
-	}
+	@ApplicationScoped
+	public static class Repository implements PanacheRepository<Settings> {
 
+		public Settings get() {
+			return Objects.requireNonNull(findById((long) SINGLETON_ID), "Settings not initialized");
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/entities/Settings.java]: fixed in https://github.com/cryptomator/hub/pull/266/commits/081349e42c5e0920a2df477b77d419b08200e849

File: backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java [hunk 6]
```diff
@@ -286,6 +296,7 @@ public void testCreateVault3() {
 		@Test
 		@Order(2)
 		@DisplayName("PUT /vaults/7E57C0DE-0000-4000-8000-000100003333 returns 200, updating only name, description and archive flag")
+		@DBRollback
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java]: No, because this is a different nested test.

I added an additional db rollback extenstion in https://github.com/cryptomator/hub/pull/266/commits/4f023e0ce15588f06fead5512294c41af2cccf77, to indicate wether the db should be reset before or after the test.

File: backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java [hunk 7]
```diff
@@ -840,19 +844,23 @@ public class ClaimOwnership {
...
+			try (var c = dataSource.getConnection(); var s = c.createStatement()) {
+				s.execute("""
+						INSERT INTO "vault" ("id", "name", "description", "creation_time", "salt", "iterations", "masterkey", "auth_pubkey", "auth_prvkey", "archived")
+						VALUES
+						('7E57C0DE-0000-4000-8000-000100009999', 'ownership-test-vault', 'Testing ownership.', '2020-02-20 20:20:20', NULL, NULL , NULL,
+						 '%s',
+						 NULL,
+						 FALSE);
+						""".formatted(authPubKey));
+			}
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java:859]: No. We need two things for persisting an object via Panache:
* the injected repository `vaultRepo`
* the `@Transactional` annotation

The former requires the method to be non-static.
The latter adds an interceptor to the method. Interceptors can only be applied on managed beans, which [inner classes are not](https://github.com/quarkusio/quarkus/issues/30789).

Hence, the usage of SQL.

File: backend/src/main/java/org/cryptomator/hub/entities/Authority.java [hunk 8]
```diff
@@ -31,21 +32,29 @@ WHERE LOWER(a.name) LIKE :name
 				FROM Authority a
 				WHERE a.id IN :ids
 				""")
-public class Authority extends PanacheEntityBase { // TODO make sealed?
+public class Authority {
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/entities/Authority.java:35]: No. Hibernate applies [lazy loading on entities by default](https://www.baeldung.com/hibernate-lazy-eager-loading#Hibernate-1). In order to do so, it uses proxy classes. But sealed classes does not allow proxies and the error
> Caused by: java.lang.IncompatibleClassChangeError: class org.cryptomator.hub.entities.Authority$HibernateProxy$Cmfc7U8x cannot inherit from sealed class org.cryptomator.hub.entities.Authority

is thrown.

File: backend/src/main/java/org/cryptomator/hub/api/VaultResource.java [hunk 9]
```diff
@@ -72,6 +71,37 @@
...
+	@Inject
+	VaultMemberAddedEvent.Repository vaultMemberAddedEventRepo;
+	@Inject
+	VaultMemberRemovedEvent.Repository vaultMemberRemovedEventRepo;
+	@Inject
+	VaultMemberUpdatedEvent.Repository vaultMemberUpdatedEventRepo;
+	@Inject
+	VaultOwnershipClaimedEvent.Repository vaultOwnershipClaimedEventRepo;
+	@Inject
+	VaultUpdatedEvent.Repository vaultUpdatedEventRepo;
```
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/api/VaultResource.java]: It is possible. I changed it in https://github.com/cryptomator/hub/pull/266/commits/75514a2e3975e24fbe9fd703563affb876cd9e4f

**Comment by infeo** [comment]
> Can we make the entities' fields private? Or what is the rationale of making them just package-private?

I changed this in da069504dd3c1c2d7e51b6bb0c58194a44d400df . Honestly, at the time of doing the migration, i was just lazy 😓 

> Shouldn't we at least for the *Event type entities try to use a single repository...

I didn't combined in a single repository, but changed the code such that all calls are bundled in a single class `EventLogger`.

File: backend/src/main/java/org/cryptomator/hub/entities/LegacyAccessToken.java [hunk 10]
```diff
@@ -119,4 +133,18 @@ public String toString() {
...
 		}
 	}
+
+	@ApplicationScoped
+	@Deprecated
+	public static class Repository implements PanacheRepositoryBase<LegacyAccessToken, AccessId> {
+
+		public LegacyAccessToken unlock(UUID vaultId, String deviceId, String userId) {
+			return find("#LegacyAccessToken.get", Map.of("deviceId", deviceId, "vaultId", vaultId, "userId", userId))
+					.firstResult();
```
Comment (overheadhunter) [inline backend/src/main/java/org/cryptomator/hub/entities/LegacyAccessToken.java:143]: isn't there always a `.singleResult()` method that throws an exception in case of >1 results? we have a primary key on `device_id, vault_id` and don't do any outer joins in this query, so an ambiguous result should be treated as errorneous.

File: backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java [hunk 11]
```diff
@@ -95,62 +97,67 @@ void applyInitialHubIdAndLicense(String initialId, String initialLicense) {
...
 			} else {
-				LOG.error("Failed to refresh license token with response code: " + response.statusCode());
+				throw new LicenseRefreshFailedException(response.statusCode(), body);
 			}
-		} catch (IOException | JWTVerificationException e) {
-			LOG.error("Failed to refresh license token", e);
 		}
 	}
 
+	//necessary for skipExecutionIf
```
Comment (overheadhunter) [inline backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java:147]: would be more legible in a public static inner class with a proper name such as `LicenseRefreshSkipper`

File: backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (overheadhunter) [reply backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java:857]: I trust you tried, but I still fail to understand this

1. the method isn't static any longer
2. `@Transactional` seemingly worked before (when this also was a nested test / inner class)

File: backend/src/main/java/org/cryptomator/hub/api/VaultResource.java [hunk 12]
```diff
@@ -443,48 +454,48 @@ public Response createOrUpdate(@PathParam("vaultId") UUID vaultId, @Valid @NotNu
...
 	@APIResponse(responseCode = "409", description = "owned by another user")
 	public Response claimOwnership(@PathParam("vaultId") UUID vaultId, @FormParam("proof") @Valid @ValidJWS String proof) {
-		User currentUser = User.findById(jwt.getSubject());
-		Vault vault = Vault.<Vault>findByIdOptional(vaultId).orElseThrow(NotFoundException::new);
+		User currentUser = userRepo.findById(jwt.getSubject());
+		Vault vault = vaultRepo.findByIdOptional(vaultId).orElseThrow(NotFoundException::new);
 
 		// if vault.authenticationPublicKey no longer exists, this vault has already been claimed by a different user
-		var authPubKey = vault.getAuthenticationPublicKey().orElseThrow(() -> new ClientErrorException(Response.Status.CONFLICT));
+		var authPubKey = vault.getAuthenticationPublicKeyOptional().orElseThrow(() -> new ClientErrorException(Response.Status.CONFLICT));
```
Comment (overheadhunter) [inline backend/src/main/java/org/cryptomator/hub/api/VaultResource.java:461]: a rather unrelated change, isn't it?

note that the `findByIdOptional()` are due to naming conflicts with `findById()`, however in this case the method does not need to carry the utility type in its name. That's like naming a method `HashMap<String, String> getFoobarHashMap()` imo.

**Review by overheadhunter — CHANGES_REQUESTED** [review]
Most things solved, just minor remarks left.

File: backend/src/main/java/org/cryptomator/hub/api/VaultResource.java (hunk 12 above)
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/api/VaultResource.java:461]: > unrelated change

No. Every entity uses the jakarta entity persistent property pattern, i.e. accessing the fields via their getter/setter. The getters and setter naming follows the [JavaBeans-style](https://jakarta.ee/learn/docs/jakartaee-tutorial/current/persist/persistence-intro/persistence-intro.html#_persistent_properties). In this case, we have the field `String authenticationPublicKey`. The according getter would be defined as `String getAuthenticationPublicKey()`. But we already have a method with this name with the same erasure.

To additionally getting a parsed key, i renamed the existing getter similar to the Quarkus naming style.

File: backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/test/java/org/cryptomator/hub/api/VaultResourceIT.java]: After a call, we found a way to [acquire an transaction context programmatically](https://quarkus.io/guides/transaction#programmatic-approach).

Changed in https://github.com/cryptomator/hub/pull/266/commits/1791f4f050f4f0580b862751715e060b6097679e.

File: backend/src/main/java/org/cryptomator/hub/entities/LegacyAccessToken.java (hunk 10 above)
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/entities/LegacyAccessToken.java]: fixed in https://github.com/cryptomator/hub/pull/266/commits/6a08760f3c06c6d4a1162287ea74fd16343e1433

File: backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java (hunk 11 above)
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/license/LicenseHolder.java]: refactored in f135a31c19b88918de69e101cb6aa08df6ee8fbf

File: backend/src/main/java/org/cryptomator/hub/entities/Authority.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (infeo) [reply backend/src/main/java/org/cryptomator/hub/entities/Authority.java:35]: > does declaring this class abstract change things?

Nope.

**Review by overheadhunter — APPROVED** [review]
